macrosynergy/panel/adjust_weights.py

```diff
@@ -0,0 +1,258 @@
+"""
+Implementation of adjust_weights.
+"""
+
+import numpy as np
+import pandas as pd
+from typing import List, Tuple, Callable, Dict, Any
+import warnings
+from numbers import Number
+from macrosynergy.management.utils import reduce_df, get_cid
+from macrosynergy.management.simulate import make_test_df
+from macrosynergy.management.types import QuantamentalDataFrame
+
+
+def check_missing_cids_xcats(weights, adj_zns, cids, r_xcats, r_cids):
+    """
+    Checks if there are missing cids or xcats in the input DataFrame.
+    """
+    missing_xcats = list(set([weights, adj_zns]) - set(r_xcats))
+    if missing_xcats:
+        raise ValueError(f"Missing xcats: {missing_xcats}")
+
+    missing_cids = list(set(cids) - set(r_cids))
+    if missing_cids:
+        raise ValueError(f"Missing cids: {missing_cids}")
+
+
+def check_types(
+    weights: str,
+    adj_zns: str,
+    method: Callable,
+    params: Dict[str, Any],
+    cids: List[str],
+):
+    """
+    Type checking for the input variables of adjust_weights.
+    """
+    for _var, _name, _type in [
+        (weights, "weights", str),
+        (adj_zns, "adj_zns", str),
+        (method, "method", Callable),
+        (params, "param", dict),
+        (cids, "cids", list),
+    ]:
+        if not isinstance(_var, _type):
+            raise TypeError(f"{_name} must be a {_type}, not {type(_var)}")
+
+    if not all(isinstance(cid, str) for cid in cids):
+        raise TypeError("`cids` must be a list of strings")
+
+
+def adjust_weights_backend(
+    df_weights_wide: pd.DataFrame,
+    df_adj_zns_wide: pd.DataFrame,
+    method: Callable,
+    params: Dict[str, Any] = {},
+) -> pd.DataFrame:
+    """
+    Backend function for adjust_weights. Applies the `method` function to the weights and
+    multiplies the result by the adjustment factors, and by the parameter `param`.
+    Expects the input DataFrames to be in wide format, with the same columns AND index
+    (see macrosynergy.panel.adjust_weights.split_weights_adj_zns).
+
+    Parameters
+    ----------
+    df_weights_wide : pd.DataFrame
+        DataFrame with weights in wide format.
+
+    df_adj_zns_wide : pd.DataFrame
+        DataFrame with adjustment factors in wide format.
+
+    method : Callable
+        Function that will be applied to the weights to adjust them.
+
+    params : Dict[str, Any], optional
+        Parameters to be passed to the method function. Default is {}.
+
+    Returns
+    -------
+    pd.DataFrame
+        DataFrame with the adjusted weights.
+    """
+
+    assert set(df_weights_wide.columns) == set(df_adj_zns_wide.columns)
+    assert set(df_weights_wide.index) == set(df_adj_zns_wide.index)
+
+    dfw_result = df_weights_wide * df_adj_zns_wide.apply(method, **params)
```

Comment:
My one worry is the use of `.apply` which to my knowledge is notoriously slow.

Could we apply something like [functools partial](https://www.learnpython.org/en/Partial_functions) to it and then use it directly?

```python
from functools import partial
fn = partial(method, **params)
dfw_result = df_weights_wide * fn(df_adj_zns_wide)
```
or for the last line `dfw_result = df_weights_wide * df_adj_zns_wide.map(fn)`.

Reply:
How is the performance of `.map(method, **params)`?

```python
dfw_result = df_weights_wide * df_adj_zns_wide.map(method, **params)
```